packages/engine/src/scene/classes/Clouds.ts

```diff
@@ -0,0 +1,210 @@
+import {
+  Mesh,
+  InstancedBufferGeometry,
+  PlaneBufferGeometry,
+  ShaderMaterial,
+  Vector3,
+  InstancedBufferAttribute,
+  Texture,
+  RawShaderMaterial,
+  UniformsUtils,
+  Fog,
+  Vector2,
+  Color
+} from 'three'
+import loadTexture from '../../editor/functions/loadTexture'
+import SimplexNoise from 'simplex-noise'
+
+const vertexShader = `
+precision highp float;
+
+attribute vec4 particles;
+attribute float particleAngle;
+
+varying vec2 vUV;
+
+void main(){
+  vUV = uv;
+
+  vec3 rotatedPosition = position;
+  rotatedPosition.x = cos( particleAngle ) * position.x - sin( particleAngle ) * position.y;
+  rotatedPosition.y = sin( particleAngle ) * position.x + cos( particleAngle ) * position.y;
+  float scale = particles.w;
+  rotatedPosition *= scale;
+
+  vec4 mvPosition = modelViewMatrix * vec4(particles.xyz, 1);
+  mvPosition.xyz += rotatedPosition;
+  gl_Position = projectionMatrix * mvPosition;
+}
+`
+const fragmentShader = `
+precision highp float;
+
+uniform sampler2D map;
+
+uniform vec3 fogColor;
+uniform float fogNear;
+uniform float fogFar;
+
+varying vec2 vUV;
+
+void main() {
+  float depth = gl_FragCoord.z / gl_FragCoord.w;
+  float fogFactor = smoothstep( fogNear, fogFar, depth );
+
+  gl_FragColor = texture2D(map,  vUV);
+  gl_FragColor.w *= pow( gl_FragCoord.z, 20.0 );
+	gl_FragColor = mix( gl_FragColor, vec4( fogColor, gl_FragColor.w ), fogFactor );
+}
+`
+
+export class Clouds extends Mesh {
+  worldScale: Vector3
+  dimensions: Vector3 // Number of particles (x,y,z)
+  noiseZoom: Vector3
+  noiseOffset: Vector3
+  spriteScaleRange: Vector2
+  fogRange: Vector2
+  fogColor: Color
+  noise: SimplexNoise
+
+  constructor(texture: Texture, configs: any) {
```

Comment:
can this please be refactored, so arguments are not applied retrospectively? instead, i suggest create the material and geometry in the constructor, but then updating the uniforms in a `setTexture` function, that can then be called each time the texture needs to be updated, rather than reaching into the uniforms directly.